File: AuthService/src/main/java/com/squareup/digital/controller/AuthController.java
package com.squareup.digital.controller;


import com.squareup.digital.dto.LoginRequest;
import com.squareup.digital.dto.LoginResponse;
import com.squareup.digital.dto.ProfileResponse;
import com.squareup.digital.model.UserModel;
import com.squareup.digital.service.AuthService;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @GetMapping
    public String index() {
        return "Welcome to the Auth API!";
    }

    @PostMapping("/register")
    public String register(@Valid @RequestBody UserModel request) {
        return authService.registerUser(request);
    }

    @PostMapping("/login")
    public LoginResponse login(@Valid @RequestBody LoginRequest request) {
        return authService.loginUser(request);
    }

    @GetMapping("/users/{username}")
    public ProfileResponse getProfile(@PathVariable String username) {
        return authService.getProfile(username);
    }
}

File: AuthService/src/main/kotlin/com/squareup/digital/AuthController.kt
package com.squareup.digital

import jakarta.validation.Valid
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/auth")
class AuthController(private val userService: UserService) {

  @GetMapping
  fun index(): String {
    return "Welcome to the Auth API!"
  }

  // 1. Standard Registration
  @PostMapping("/register")
  fun register(@Valid @RequestBody request: UserModel): String {
    userService.registerUser(request)
    return "User registered successfully"
  }

  // 2. Standard Login
  @PostMapping("/login")
  fun login(@Valid @RequestBody request: LoginRequest): LoginResponse {
    return userService.loginUser(request)
  }

  // 3. Google OAuth2 Login
  @GetMapping("/google/success")
  fun googleLogin(@AuthenticationPrincipal principal: OAuth2User): LoginResponse {
    return userService.processGoogleLogin(principal)
  }
}
